atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/cache/ValueStoreImplTest.java

```diff
@@ -0,0 +1,145 @@
+        assertExpectedValue(CELL_2, CacheEntry.unlocked(VALUE_3));
+
+        valueStore.applyEvent(LOCK_EVENT);
+        assertExpectedValue(CELL_1, CacheEntry.locked());
+        assertExpectedValue(CELL_2, CacheEntry.unlocked(VALUE_3));
+    }
+
+    @Test
+    public void unlockEventsClearLockedEntries() {
+        valueStore.applyEvent(WATCH_EVENT);
+        valueStore.applyEvent(LOCK_EVENT);
+
+        assertExpectedValue(CELL_1, CacheEntry.locked());
+
+        valueStore.applyEvent(UNLOCK_EVENT);
+        assertThat(valueStore.getSnapshot().getValue(TABLE_CELL)).isEmpty();
+    }
+
+    @Test
+    public void putValueThrowsIfCurrentValueDiffers() {
+        valueStore.applyEvent(WATCH_EVENT);
+        valueStore.putValue(TABLE_CELL, VALUE_1);
+
+        assertThatCode(() -> valueStore.putValue(TABLE_CELL, VALUE_1)).doesNotThrowAnyException();
+        assertPutThrows(VALUE_2);
+
+        valueStore.applyEvent(LOCK_EVENT);
+        assertPutThrows(VALUE_1);
+    }
+
+    @Test
+    public void watchEventUpdatesWatchableTables() {
+        assertThat(valueStore.getSnapshot().isWatched(TABLE)).isFalse();
+        valueStore.applyEvent(WATCH_EVENT);
+        assertThat(valueStore.getSnapshot().isWatched(TABLE)).isTrue();
+    }
+
+    private void assertPutThrows(CacheValue value) {
+        assertThatThrownBy(() -> valueStore.putValue(TABLE_CELL, value))
+                .isExactlyInstanceOf(SafeIllegalStateException.class)
+                .hasMessageContaining(
+                        "Trying to cache a value which is either locked or is not equal to a currently cached value");
+    }
+
+    private void assertExpectedValue(Cell cell, CacheEntry entry) {
+        assertThat(valueStore.getSnapshot().getValue(CellReference.of(TABLE, cell)))
+                .hasValue(entry);
+    }
+
+    private static LockWatchEvent createWatchEvent() {
+        return LockWatchCreatedEvent.builder(
+                        ImmutableSet.of(LockWatchReferences.entireTable(TABLE.getQualifiedName())), ImmutableSet.of())
+                .build(0L);
+    }
+
+    private static LockWatchEvent createLockEvent() {
+        return LockEvent.builder(
+                        ImmutableSet.of(AtlasCellLockDescriptor.of(
+                                TABLE.getQualifiedName(), CELL_1.getRowName(), CELL_1.getColumnName())),
+                        LockToken.of(UUID.randomUUID()))
+                .build(1L);
+    }
+
+    private static LockWatchEvent createUnlockEvent() {
+        return UnlockEvent.builder(ImmutableSet.of(AtlasCellLockDescriptor.of(
+                        TABLE.getQualifiedName(), CELL_1.getRowName(), CELL_1.getColumnName())))
+                .build(1L);
+    }
+
+    private static CacheValue createValue(int value) {
+        return CacheValue.of(createBytes(value));
+    }
+
+    private static Cell createCell(int value) {
+        return Cell.create(createBytes(value), createBytes(value + 100));
```

Comment:
non-actionable: I wonder when someone will call `createBytes(28)`...

Reply:
It's actually fine - before I was using `value + 999`, and it just removes the excess. I know +28 is within a byte's abilities but is negative, but I believe cells are unsigned, right? (Or that the sign doesn't really matter).